src/Test.UnitTests.Sarif.WorkItems/SarifWorkItemFilerTests.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.WorkItems;
+using Moq;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Sarif.WorkItems
+{
+    public class SarifWorkItemFilerTests
+    {
+        private static readonly Uri s_testUri = new Uri("https://github.com/Microsoft/sarif-sdk");
+
+        [Fact]
+        public void WorkItemFiler_ValidatesSarifLogFileContentsArgument()
+        {
+            var filer = CreateWorkItemFiler();
+
+            Action action = () => filer.FileWorkItems(sarifLogFileContents: null);
+
+            action.Should().Throw<ArgumentNullException>();
+        }
+
+        [Fact]
+        public void WorkItemFiler_ValidatesSarifLogFileArgument()
+        {
+            var filer = CreateWorkItemFiler();
+
+            Action action = () => filer.FileWorkItems(sarifLog: null);
+
+            action.Should().Throw<ArgumentNullException>();
+        }
+
+        [Fact]
+        public void WorkItemFiler_ValidatesSarifLogFileLocationArgument()
+        {
+            var filer = CreateWorkItemFiler();
+
+            Action action = () => filer.FileWorkItems(sarifLogFileLocation: null);
+
+            action.Should().Throw<ArgumentNullException>();
+        }
+
+        private static SarifWorkItemFiler CreateWorkItemFiler()
+            => CreateMockSarifWorkItemFiler();
+
+        private static SarifWorkItemFiler CreateMockSarifWorkItemFiler()
+        {
+            var mockFilingTarget = new Mock<SarifWorkItemFiler>();
+
+            mockFilingTarget
+                .Setup(x => x.FileWorkItems(It.IsAny<string>()))
+                .CallBase(); 
+
+            mockFilingTarget
+                .Setup(x => x.FileWorkItems(It.IsAny<SarifLog>()))
+                .CallBase();
+
+            mockFilingTarget
+                .Setup(x => x.FileWorkItems(It.IsAny<Uri>()))
+                .CallBase();
+            
+            // Moq magic: you can return whatever was passed to a method by providing
+            // a lambda (rather than a fixed value) to Returns or ReturnsAsync.
+            // https://stackoverflow.com/questions/996602/returning-value-that-was-passed-into-a-method
+            mockFilingTarget
+                .Setup(x => x.FileWorkItems(
+                    It.IsAny<Uri>(), 
+                    It.IsAny<IList<WorkItemModel<SarifWorkItemContext>>>(),
+                    It.IsAny<string>()))
+                .ReturnsAsync((Uri uri, IList<WorkItemModel<SarifWorkItemContext>> resultGroups, string securityToken) => resultGroups);
+
+            return mockFilingTarget.Object;
+        }
+
+        private class TestFilingClient : FilingClient
```

Comment:
>TestFilingClient [](start = 22, length = 16)

I'm curious about the mixture of use of Moq on the one hand vs. use of fake objects like this on the other. You can set up a Moq to throw an exception.

Reply:
If you use Moq exclusively, then you miss the opportunity to observe what it looks like to provide actual extensions. i.e., if you have an extensibility model, you may as well use it for injecting things, as it exercises more design surface area. That is my thinking.

---
In reply to: [381487051](https://github.com/microsoft/sarif-sdk/pull/1775#discussion_r381487051) [](ancestors = 381487051)